使用前に  
・dateカラムとidカラムは「セルの書式設定」で「文字列」にする（議案データであればB, D, N列）  
・dateカラムが４桁かつ月日がバラバラになっていないか確認  
　　例）123は1月23日の「0123」なのか12月3日の「1203」なのか分からないため  
・編集したいファイルを予めGoogleドライブに保存

１．はじめに Google ドライブのファイルへのアクセスを許可する（左のドライブファイルからGoogle ドライブをマウントする）

In [33]:
!pip install openpyxl

２．編集したいファイルパスを指定（左のフォルダから編集したいファイルを副ボタンでクリック。「パスをコピー」し、' 'の中に貼り付け）  
３．保存先のファイルパスと好きなファイル名を指定

In [34]:
import openpyxl

import_path = r'/content/drive/My Drive/iinkai_test.xlsx' # 編集したいファイルのパスを作成
wb = openpyxl.load_workbook(import_path, data_only=True)
ws = wb['Sheet1'] # 編集したいシート名
export_path = r'/content/drive/My Drive/iinkai_result.xlsx' # 保存先のファイルパスとファイル名を指定

４．IDを作成したいデータのプログラムを実行（議案、委員会）

議案データ

In [31]:
def is_empty(cell):
  return cell.value is None or not str(cell.value).strip()

i = 2
max = ws.max_column
for row in ws.iter_rows(min_row=2):
    # 空行になったら終了
    if all(is_empty(c) for c in row):
      break
    # 下３桁の数字リセットサイン
    date_cell = str(ws.cell(row=i, column=2).value)
    pre_date_cell = str(ws.cell(row=i-1, column=2).value)
    if date_cell == pre_date_cell:
      ws.cell(row=i, column=max+1).value = 1
    else:
      ws.cell(row=i, column=max+1).value = 0
    # 下３桁の数字作成
    if ws.cell(row=i, column=max+1).value == 0:
      ws.cell(row=i, column=max+2).value = 1
    else:
      ws.cell(row=i, column=max+2).value = ws.cell(row=i-1, column=max+2).value + 1
    # id作成
    year = str(ws.cell(row=i, column=1).value)
    date = str(ws.cell(row=i, column=2).value).zfill(4)
    jis = str(ws.cell(row=i, column=3).value)
    num = str(ws.cell(row=i, column=max+2).value).zfill(3)
    id = year + date + jis + num
    ws.cell(row=i, column=4).value = str(id)
    ws.cell(row=i, column=max+3).value = str(id)
    i+=1

    values = []
    for col in row:
      values.append(col.value)
    print(values)
    wb.save(export_path)

[2014, '0228', 29201, '2014022829201001', '市長専決処分の報告について', None, None, 0, 1, '報告', '報告', None, None, None]
[2014, '0228', 29201, '2014022829201002', '市長専決処分の報告について', None, None, 0, 2, '報告', '報告', None, None, None]
[2014, '0228', 29201, '2014022829201003', '市長専決処分の報告について', None, None, 0, 3, '報告', '報告', None, None, None]
[2014, '0228', 29201, '2014022829201004', '市長専決処分の報告について', None, None, 0, 4, '報告', '報告', None, None, None]
[2014, '0228', 29201, '2014022829201005', '市長専決処分の報告について', None, None, 0, 5, '報告', '報告', None, None, None]
[2014, '0228', 29201, '2014022829201006', '市長専決処分の報告について', None, None, 0, 6, '報告', '報告', None, None, None]
[2014, '0228', 29201, '2014022829201007', '市長専決処分の報告について', None, None, 0, 7, '報告', '報告', None, None, None]
[2014, '0228', 29201, '2014022829201008', '株式会社奈良市清美公社の事業計画の報告について', None, None, 0, 8, '報告', '報告', None, None, None]
[2014, '0228', 29201, '2014022829201009', '奈良市市街地開発株式会社の事業計画の報告について', None, None, 0, 9, '報告', '報告', None, None, None]
[2014, '0228', 2

委員会データ

４-1．ID作成したい市町村コードを下記より指定  

自治体コード	団体名
  
滋賀県  
25201	大津市　25202	 彦根市　25203	長浜市　25204	近江八幡市	  
25206	草津市　25207	守山市　25208	栗東市　25209	甲賀市	  
25210	野洲市　25211	湖南市　25212	高島市　25213	東近江市	  
25214	米原市	  

奈良県  
29201	奈良市　29202	大和高田市　29203	大和郡山市　29204	天理市	  
29205	橿原市　29206	桜井市　29207	五條市　29208	御所市  
29209	生駒市　29210	香芝市　29211	葛城市　29212	宇陀市	  

和歌山県  
30201	和歌山市　30202	海南市　30203	橋本市　30204	有田市	  
30205	御坊市　30206	田辺市　30207	新宮市　30208	紀の川市	  
30209	岩出市  

In [35]:
code = 30202 #市町村コードの指定

In [36]:
def is_empty(cell):
  return cell.value is None or not str(cell.value).strip()

i = 2
max = ws.max_column
for row in ws.iter_rows(min_row=2):
    # 空行になったら終了
    if all(is_empty(c) for c in row):
      break
    # 下３桁の数字リセットサイン
    date_cell = str(ws.cell(row=i, column=2).value)
    pre_date_cell = str(ws.cell(row=i-1, column=2).value)
    if date_cell == pre_date_cell:
      ws.cell(row=i, column=max+1).value = 1
    else:
      ws.cell(row=i, column=max+1).value = 0
    # 下３桁の数字作成
    if ws.cell(row=i, column=max+1).value == 0:
      ws.cell(row=i, column=max+2).value = 1
    else:
      ws.cell(row=i, column=max+2).value = ws.cell(row=i-1, column=max+2).value + 1
    # id作成
    date = str(ws.cell(row=i, column=2).value)
    jis = str(code)
    num = str(ws.cell(row=i, column=max+2).value).zfill(3)
    id = date + jis + num
    ws.cell(row=i, column=3).value = str(id)
    ws.cell(row=i, column=max+3).value = str(id)
    i+=1

    values = []
    for col in row:
      values.append(col.value)
    print(values)
    wb.save(export_path)

[20100519, 20100425, '2010042530202001', '海南市', 2344, '中西\u3000徹', '中西\u3000徹', 'ナカニシ\u3000トオル', 37, '男', '市議会議員', '現', '無', None, None, None, None, None, 3, None, None, None, None, 3, None, None, None, None, None, None, None, None, 1]
[20100519, 20100425, '2010042530202002', '海南市', 1602, '宮本\u3000勝利', '宮本\u3000勝利', 'ミヤモト\u3000カツトシ', 68, '男', '無職', '現', '無', None, None, None, 2, None, None, None, None, None, None, 3, None, None, None, None, None, None, None, None, 3]
[20100519, 20100425, '2010042530202003', '海南市', 1494, '中家\u3000悦生', '中家\u3000悦生', 'ナカヤ\u3000エツオ', 48, '男', '団体役員', '現', '公', None, None, None, None, 3, None, None, None, None, None, 2, None, None, None, None, None, None, None, 1, None]
[20100519, 20100425, '2010042530202004', '海南市', 1436, '出口\u3000茂治', '出口\u3000茂治', 'デグチシゲハル', 66, '男', '石油会社員', '現', '無', None, None, None, None, 3, None, None, None, None, None, 3, None, None, None, None, None, None, None, 3, None]
[20100519, 20100425, '2010042530202005', '海南市', 1429, '前山\u3